build.assets/tooling/cmd/buf-plugin-linters/main.go

```diff
@@ -0,0 +1,169 @@
+// Teleport
+// Copyright (C) 2025 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+package main
+
+import (
+	"context"
+	"strings"
+
+	"buf.build/go/bufplugin/check"
+	"buf.build/go/bufplugin/check/checkutil"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+const (
+	paginationRuleID = "PAGINATION_REQUIRED"
+)
+
+var paginationSpec = &check.Spec{
+	Rules: []*check.RuleSpec{
+		{
+			ID:      paginationRuleID,
+			Purpose: "Ensure RPCs returning a repeated field use pagination fields.",
+			Type:    check.RuleTypeLint,
+			Handler: checkutil.NewMethodRuleHandler(checkPaginationMethod),
+		},
+	},
+}
+
+func main() {
+	check.Main(paginationSpec)
+}
+
+// checkPaginationMethod implements MethodRuleHandler for RuleSpec
+func checkPaginationMethod(
+	_ context.Context,
+	responseWriter check.ResponseWriter,
+	request check.Request,
+	method protoreflect.MethodDescriptor,
+) error {
+
+	if method.IsStreamingServer() || method.IsStreamingClient() {
+		// Streaming APIs do not expect pagination.
+		return nil
+	}
+
+	if isMethodDeprecated(method) {
+		// deprecated methods are skipped
+		return nil
+	}
+
+	resp := method.Output()
+	if !hasRepeated(resp.Fields()) {
+		// Check if the method *should* have the repeated field.
+		if strings.HasPrefix(string(method.Name()), methodPrefixMustHaveRepeated) {
+			responseWriter.AddAnnotation(
+				check.WithDescriptor(method),
+				check.WithMessagef(
+					"repeated fields expected for RPC names starting with: %q (RFD-0153)",
+					methodPrefixMustHaveRepeated,
+				),
+			)
+		}
+		return nil
+	}
+
+	config := newDefaultConfig()
+	if config.shouldSkip(method) {
+		return nil
+	}
```

Comment:
Move the skip before the check for repeated fields? I don't know how much effort it takes to calculate and check the method descriptor but we have a lot of methods, it might be a little faster to exclude methods by name immediately.

Reply:
Ignore the performance comment, move the skip check before everything else because otherwise we won't be able to skip methods that fail the previous check 😅